drivers/sensor/bme680/bme680.c

```diff
@@ -0,0 +1,427 @@
+	heatr_res = (heatr_res_x100 + 50) / 100;
+
+	return heatr_res;
+}
+
+static u8_t bme680_calc_gas_wait(u16_t dur)
+{
+	u8_t factor = 0, durval;
+
+	if (dur >= 0xfc0) {
+		durval = 0xff; /* Max duration*/
+	} else {
+		while (dur > 0x3F) {
+			dur = dur / 4;
+			factor += 1;
+		}
+		durval = dur + (factor * 64);
+	}
+
+	return durval;
+}
+
+static int bme680_sample_fetch(struct device *dev, enum sensor_channel chan)
+{
+	struct bme680_data *data = dev->driver_data;
+	u8_t buff[BME680_LEN_FIELD] = { 0 };
+	u8_t gas_range;
+	u32_t adc_temp, adc_press;
+	u16_t adc_hum, adc_gas_res;
+	int size = BME680_LEN_FIELD;
+	int ret;
+
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_ALL);
+
+	ret = bme680_reg_read(data, BME680_REG_FIELD0, buff, size);
+	if (ret < 0) {
+		return ret;
+	}
+
+	data->new_data = buff[0] & BME680_MSK_NEW_DATA;
+	data->heatr_stab = buff[14] & BME680_MSK_HEATR_STAB;
+
+	adc_press = (u32_t)(((u32_t)buff[2] * 4096) | ((u32_t)buff[3] * 16)
+			    | ((u32_t)buff[4] / 16));
+	adc_temp = (u32_t)(((u32_t)buff[5] * 4096) | ((u32_t)buff[6] * 16)
+			   | ((u32_t)buff[7] / 16));
+	adc_hum = (u16_t)(((u32_t)buff[8] * 256) | (u32_t)buff[9]);
+	adc_gas_res = (u16_t)((u32_t)buff[13] * 4 | (((u32_t)buff[14]) / 64));
+	gas_range = buff[14] & BME680_MSK_GAS_RANGE;
+
+	if (data->new_data) {
+		bme680_calc_temp(data, adc_temp);
+		bme680_calc_press(data, adc_press);
+		bme680_calc_humidity(data, adc_hum);
+		bme680_calc_gas_resistance(data, gas_range, adc_gas_res);
+	}
+
+	/* Trigger the next measurement */
+	ret = bme680_reg_write(data, BME680_REG_CTRL_MEAS,
+			       BME680_CTRL_MEAS_VAL);
+	if (ret < 0) {
+		return ret;
+	}
+
+	return 0;
+}
+
+static int bme680_channel_get(struct device *dev, enum sensor_channel chan,
+			      struct sensor_value *val)
+{
+	struct bme680_data *data = dev->driver_data;
+
+	switch (chan) {
+	case SENSOR_CHAN_AMBIENT_TEMP:
+		/*
+		 * data->calc_temp has a resolution of 0.01 degC.
+		 * So 5123 equals 51.23 degC.
+		 */
+		val->val1 = data->calc_temp / 100;
+		val->val2 = data->calc_temp % 100 * 10000;
+		break;
+	case SENSOR_CHAN_PRESS:
+		/*
+		 * data->calc_press has a resolution of 1 Pa.
+		 * So 96321 equals 96.321 kPa.
+		 */
+		val->val1 = data->calc_press / 1000;
+		val->val2 = (data->calc_press % 1000) * 1000;
+		break;
+	case SENSOR_CHAN_HUMIDITY:
+		/*
+		 * data->calc_humidity has a resolution of 0.001 %RH.
+		 * So 46333 equals 46.333 %RH.
+		 */
+		val->val1 = data->calc_humidity / 1000;
+		val->val2 = (data->calc_humidity % 1000) * 1000;
+		break;
+	case SENSOR_CHAN_GAS_RES:
+		/*
+		 * data->calc_gas_resistance has a resolution of 1 ohm.
+		 * So 100000 equals 100000 ohms.
+		 */
+		val->val1 = data->calc_gas_resistance;
+		val->val2 = 0;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int bme680_read_compensation(struct bme680_data *data)
+{
+	u8_t buff[BME680_LEN_COEFF_ALL];
+	int err = 0;
+
+	err = bme680_reg_read(data, BME680_REG_COEFF1, buff, BME680_LEN_COEFF1);
```

Comment:
Is it not possible to do one burst read instead of three separate transactions? Does the sensor support register autoincrement?